certora/specs/RevertsERC20.spec

```
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+methods {
+    function totalSupply() external returns uint256 envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function allowance(address, address) external returns uint256 envfree;
+    function delegatee(address) external returns address envfree;
+    function delegatedVotingPower(address) external returns uint256 envfree;
+}
+
+// Check the revert conditions for the transfer function.
+rule transferRevertConditions(env e, address to, uint256 amount) {
+    uint256 balanceOfSenderBefore = balanceOf(e.msg.sender);
+    uint256 senderVotingPowerBefore = delegatedVotingPower(delegatee(e.msg.sender));
+    uint256 recipientVotingPowerBefore = delegatedVotingPower(delegatee(to));
+
+    // Safe require as it is verified in delegatedLTEqDelegateeVP.
+    require senderVotingPowerBefore >= balanceOfSenderBefore;
```

can we do this when it's possible? This has the good property of not having to "trust the require"
```suggestion
    requireInvariant delegatedLTEqDelegateeVP();
```

---

We can't require these invariants. The limitation stem from the fact that these invariants require a munged setup that is incompatible with these rules. Using the munged setup here needs to add new requirements that are "unsafe", see this [discussion](https://github.com/morpho-org/morpho-token/pull/98#discussion_r1873793283).

---

Agree that it's the best if we can do it. Relevant discussions: [here](https://github.com/morpho-org/morpho-token/pull/98#discussion_r1876004724) and [here](https://github.com/morpho-org/morpho-token/pull/98#discussion_r1876018666).

---

so there is nothing simple we can do here?

---

To actually use the invariants that we have proven? Not that I know of.